File: src/frontend/inventario/lib/app/data/model/inventario_ajuste_cab_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:inventario/app/infra/infra_imports.dart';
import 'package:inventario/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:inventario/app/data/domain/domain_imports.dart';

class InventarioAjusteCabModel {
	int? id;
	int? idViewPessoaColaborador;
	DateTime? dataAjuste;
	String? tipo;
	double? taxa;
	String? justificativa;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;
	List<InventarioAjusteDetModel>? inventarioAjusteDetModelList;

	InventarioAjusteCabModel({
		this.id,
		this.idViewPessoaColaborador,
		this.dataAjuste,
		this.tipo,
		this.taxa,
		this.justificativa,
		this.viewPessoaColaboradorModel,
		this.inventarioAjusteDetModelList,
	});

	static List<String> dbColumns = <String>[
		'id',
		'data_ajuste',
		'tipo',
		'taxa',
		'justificativa',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Data Ajuste',
		'Tipo',
		'Taxa',
		'Justificativa',
	];

	InventarioAjusteCabModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idViewPessoaColaborador = jsonData['idViewPessoaColaborador'];
		dataAjuste = jsonData['dataAjuste'] != null ? DateTime.tryParse(jsonData['dataAjuste']) : null;
		tipo = InventarioAjusteCabDomain.getTipo(jsonData['tipo']);
		taxa = jsonData['taxa']?.toDouble();
		justificativa = jsonData['justificativa'];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
		inventarioAjusteDetModelList = (jsonData['inventarioAjusteDetModelList'] as Iterable?)?.map((m) => InventarioAjusteDetModel.fromJson(m)).toList() ?? [];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idViewPessoaColaborador'] = idViewPessoaColaborador != 0 ? idViewPessoaColaborador : null;
		jsonData['dataAjuste'] = dataAjuste != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataAjuste!) : null;
		jsonData['tipo'] = InventarioAjusteCabDomain.setTipo(tipo);
		jsonData['taxa'] = taxa;
		jsonData['justificativa'] = justificativa;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
		
		var inventarioAjusteDetModelLocalList = []; 
		for (InventarioAjusteDetModel object in inventarioAjusteDetModelList ?? []) { 
			inventarioAjusteDetModelLocalList.add(object.toJson); 
		}
		jsonData['inventarioAjusteDetModelList'] = inventarioAjusteDetModelLocalList;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idViewPessoaColaborador = plutoRow.cells['idViewPessoaColaborador']?.value;
		dataAjuste = Util.stringToDate(plutoRow.cells['dataAjuste']?.value);
		tipo = plutoRow.cells['tipo']?.value != '' ? plutoRow.cells['tipo']?.value : 'Aumentar';
		taxa = plutoRow.cells['taxa']?.value?.toDouble();
		justificativa = plutoRow.cells['justificativa']?.value;
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
		inventarioAjusteDetModelList = [];
	}	

	InventarioAjusteCabModel clone() {
		return InventarioAjusteCabModel(
			id: id,
			idViewPessoaColaborador: idViewPessoaColaborador,
			dataAjuste: dataAjuste,
			tipo: tipo,
			taxa: taxa,
			justificativa: justificativa,
			inventarioAjusteDetModelList: inventarioAjusteDetModelListClone(inventarioAjusteDetModelList!),
		);			
	}

	inventarioAjusteDetModelListClone(List<InventarioAjusteDetModel> inventarioAjusteDetModelList) { 
		List<InventarioAjusteDetModel> resultList = [];
		for (var inventarioAjusteDetModel in inventarioAjusteDetModelList) {
			resultList.add(
				InventarioAjusteDetModel(
					id: inventarioAjusteDetModel.id,
					idInventarioAjusteCab: inventarioAjusteDetModel.idInventarioAjusteCab,
					idProduto: inventarioAjusteDetModel.idProduto,
					valorOriginal: inventarioAjusteDetModel.valorOriginal,
					valorReajuste: inventarioAjusteDetModel.valorReajuste,
				)
			);
		}
		return resultList;
	}

	
}

File: backend/inventario/src/model/inventario_ajuste_cab_model.py
from src import db
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel
from src.model.inventario_ajuste_det_model import InventarioAjusteDetModel


class InventarioAjusteCabModel(db.Model):
    __tablename__ = 'inventario_ajuste_cab'

    id = db.Column(db.Integer, primary_key=True)
    data_ajuste = db.Column(db.DateTime)
    tipo = db.Column(db.String(1))
    taxa = db.Column(db.Float)
    justificativa = db.Column(db.Text)
    id_view_pessoa_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))

    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_view_pessoa_colaborador])
    inventario_ajuste_det_model_list = db.relationship('InventarioAjusteDetModel', lazy='dynamic')

    def mapping(self, data):
        self.id = data.get('id')
        self.id_view_pessoa_colaborador = data.get('idViewPessoaColaborador')
        self.data_ajuste = data.get('dataAjuste')
        self.tipo = data.get('tipo')
        self.taxa = data.get('taxa')
        self.justificativa = data.get('justificativa')

    def serialize(self):
        return {
            'id': self.id,
            'idViewPessoaColaborador': self.id_view_pessoa_colaborador,
            'dataAjuste': self.data_ajuste.isoformat(),
            'tipo': self.tipo,
            'taxa': self.taxa,
            'justificativa': self.justificativa,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
            'inventarioAjusteDetModelList': [inventario_ajuste_det_model.serialize() for inventario_ajuste_det_model in self.inventario_ajuste_det_model_list],
        }